File: tools/bundle-preview.py
#!/usr/bin/env python3
"""
Bundle a built client site into ONE self-contained HTML page for preview.

Why: a client (or you) needs to click round the real thing before launch, and
hosting a 20+ page static site just to show someone is friction. This inlines
every page, hoists the CSS once, and swaps internal links for hash routing, so
the whole site travels as a single file you can publish or email.

It is a PREVIEW harness, not the deliverable. Page <script> blocks are stripped
and replaced with minimal routing, so JS behaviour is approximated, not real.
Always test the actual build before launch.

Usage:
    cd clients/<name> && npm run build
    python3 ../../tools/bundle-preview.py            # -> /tmp/<name>-preview.html
    python3 ../../tools/bundle-preview.py out.html
"""
import re
import sys
import json
import html
import base64
import pathlib

# An optional landing route. Without this a multi-route preview always opens on
# "/", which sent a client to an old page and made finished work look unchanged.
argv = [a for a in sys.argv[1:]]


def norm_route(r):
    if not r.startswith('/'):
        r = '/' + r
    if not r.endswith('/'):
        r += '/'
    return r


default_route = '/'
only = []
keep_scripts = False
for a in list(argv):
    if a.startswith('--default='):
        default_route = norm_route(a.split('=', 1)[1])
        argv.remove(a)
    elif a.startswith('--only='):
        only.append(norm_route(a.split('=', 1)[1]))
        argv.remove(a)
    elif a == '--scripts':
        keep_scripts = True
        argv.remove(a)

dist = pathlib.Path('dist')
css  = "\n".join(p.read_text() for p in sorted((dist/'_astro').glob('*.css')))

# Map every built page to a route
pages = {}
for f in sorted(dist.rglob('index.html')):
    route = '/' + str(f.parent.relative_to(dist)).replace('.', '').strip('/')
    route = '/' if route in ('/', '') else route.rstrip('/') + '/'
    pages[route] = f.read_text()

if not pages:
    sys.exit("No built pages found. Run `npm run build` first.")

if only:
    missing = [r for r in only if r not in pages]
    if missing:
        sys.exit(f"No such route(s): {', '.join(missing)}. Built: {', '.join(pages)}")
    pages = {r: pages[r] for r in only}
    if default_route not in pages:
        default_route = only[0]

# -----------------------------------------------------------------------------
# Scripts are stripped by default and that is the right default: the routing
# below is hash based, so keeping every route's scripts in one document runs all
# of them at once against a DOM where only one route is visible.
#
# It is also how a finished animated hero came out looking static. The Stacked
# Out calendar is a GSAP timeline over a canvas, so with scripts gone the preview
# showed the settled sheet and the page turn was simply absent — the preview was
# lying about the build in the direction that matters most.
#
# So --scripts keeps them, and it requires exactly one route, because that is the
# only case where "run the page's scripts" means what it says.
# -----------------------------------------------------------------------------
if keep_scripts and len(pages) != 1:
    sys.exit(
        "--scripts needs exactly one route, or every route's scripts run at once.\n"
        f"Add --only=<route>. Built: {', '.join(pages)}"
    )

# -----------------------------------------------------------------------------
# CRITICAL: the design tokens live on the <html> element, not in the stylesheet.
#   <html data-mode="dark" style="--brand:#22c8ee;...">
# The wrapper page supplies its own <html>, so those are lost unless they are
# lifted out explicitly. Miss this and the site renders with no palette and a
# fallback accent colour — which looks like a completely different, broken site.
# -----------------------------------------------------------------------------
home = pages.get('/') or next(iter(pages.values()))
html_tag = re.search(r'<html([^>]*)>', home)
attrs = html_tag.group(1) if html_tag else ''

inline_vars = ''
m = re.search(r'style="([^"]*)"', attrs)
if m:
    inline_vars = html.unescape(m.group(1))

mode = ''
m = re.search(r'data-mode="([^"]*)"', attrs)
if m:
    mode = m.group(1)

# Re-point the mode-scoped token blocks at :root so they apply without <html>.
# Minified CSS drops the attribute quotes ([data-mode=dark]), so match all forms.
if mode:
    css = re.sub(r"\[data-mode=['\"]?" + re.escape(mode) + r"['\"]?\]", ':root', css)
    # Drop the other mode's block entirely so it can't win on source order —
    # otherwise dark-mode text keeps the light-mode colour and vanishes.
    other = 'light' if mode == 'dark' else 'dark'
    css = re.sub(r"\[data-mode=['\"]?" + other + r"['\"]?\]\s*\{[^}]*\}", '', css)
    # Any remaining rules scoped to the other mode (e.g. `[data-mode=light] .hero`)
    # are dead weight in a single-mode preview.
    css = re.sub(r"\[data-mode=['\"]?" + other + r"['\"]?\][^{]*\{[^}]*\}", '', css)

# Fonts are same-origin files in the real build; inline them so the single file
# is genuinely self-contained (the artifact CSP blocks external font hosts too).
def inline_fonts(sheet):
    def sub(m):
        url = m.group(1).strip('\'"')
        f = dist / url.lstrip('/')
        if not f.exists():
            return m.group(0)
        b64 = base64.b64encode(f.read_bytes()).decode()
        return f"url('data:font/woff2;base64,{b64}')"
    return re.sub(r"url\((['\"]?[^)]+\.woff2['\"]?)\)", sub, sheet)

css = inline_fonts(css)

# Same problem as the fonts, and it was silently shipping broken previews: an
# <img src="/img/…"> is same-origin in the real build, but in a single published
# file that root-relative path resolves against the artifact host and 404s, so
# every photograph on the page came out as an empty box. The site looked worse in
# preview than it does in the build, which is the one thing a preview must never
# do. Raster and SVG both, keyed off the real extension so the data URI carries
# the right MIME type.
MIME = {
    '.webp': 'image/webp', '.png': 'image/png', '.jpg': 'image/jpeg',
    '.jpeg': 'image/jpeg', '.svg': 'image/svg+xml', '.avif': 'image/avif',
    '.gif': 'image/gif', '.ico': 'image/x-icon',
}

def inline_images(markup):
    def sub(m):
        attr, quote, url = m.group(1), m.group(2), m.group(3)
        f = dist / url.lstrip('/')
        mime = MIME.get(f.suffix.lower())
        if not mime or not f.exists():
            return m.group(0)
        b64 = base64.b64encode(f.read_bytes()).decode()
        return f'{attr}={quote}data:{mime};base64,{b64}{quote}'
    # src and href, and the leading slash is required: a data: or https: URL is
    # already self-contained and must be left alone.
    return re.sub(r'\b(src|href)=(["\'])(/[^"\']+)\2', sub, markup)

def inline_scripts(markup):
    """Turn <script src="/vendor/gsap.min.js"> into the code itself.

    Same reason as the fonts and the images: a root relative src 404s out of a
    single published file, and a hero whose animation library failed to load is
    worse than one with no animation at all, because the markup is then sitting
    in whatever state the timeline was meant to move it out of. defer is dropped
    deliberately — these end up inline at the point they already occupied, in
    document order, and the calendar's own script runs after them.
    """
    def sub(m):
        url = m.group(1)
        f = dist / url.lstrip('/')
        if not f.exists():
            return m.group(0)
        return '<script>' + f.read_text() + '</script>'
    return re.sub(r'<script[^>]*\bsrc=["\'](/[^"\']+\.js)["\'][^>]*>\s*</script>', sub, markup)


def body_of(doc):
    m = re.search(r'<body[^>]*>(.*)</body>', doc, re.S)
    inner = m.group(1) if m else doc
    if keep_scripts:
        inner = inline_scripts(inner)
    else:
        inner = re.sub(r'<script(?![^>]*application/ld\+json).*?</script>', '', inner, flags=re.S)
    inner = re.sub(r'<script type="application/ld\+json".*?</script>', '', inner, flags=re.S)
    inner = inline_images(inner)
    return inner

# Per-page <style> blocks (Astro scoped styles) get hoisted once
scoped = set()
for doc in pages.values():
    for m in re.finditer(r'<style>(.*?)</style>', doc, re.S):
        scoped.add(m.group(1))

sections = []
for route, doc in pages.items():
    b = body_of(doc)
    b = re.sub(r'<style>.*?</style>', '', b, flags=re.S)
    sections.append(f'<div class="rt" data-route="{html.escape(route)}" hidden>{b}</div>')

# Site title, taken from the real build rather than hard-coded.
tm = re.search(r'<title>(.*?)</title>', home, re.S)
title = html.escape(re.sub(r'\s+', ' ', tm.group(1)).strip()) if tm else 'Site preview'

out = f"""<title>{title} — preview</title>
<style>
{css}
{"".join(scoped)}
/* Tokens lifted off the built <html> element. MUST come after the stylesheet:
   global.css declares its own :root defaults, and at equal specificity the
   later rule wins. Emit these first and the client's accent silently loses to
   the template default. */
:root {{ {inline_vars} }}
/* preview shell.
   The background falls through several token names because different client
   projects name their ground differently (--bg on one, --ground on another).
   Hardcoding one project's fallback here silently repaints another brand's
   page in the wrong colour, which is exactly what it did once. */
html,body{{margin:0;background:var(--bg,var(--ground,var(--surface,#0c0d10)))}}
.rt[hidden]{{display:none}}
.pv-note{{position:fixed;left:0;right:0;bottom:0;z-index:999;background:#111318;color:#a1a5ad;
  border-top:1px solid rgba(255,255,255,.12);font:500 12px/1.45 system-ui,sans-serif;
  padding:9px 16px;text-align:center}}
.pv-note b{{color:var(--brand,var(--incoming,#22c8ee))}}
body{{padding-bottom:42px}}
</style>

{''.join(sections)}

<div class="pv-note">
  Preview of a real {len(pages)}-route build. Dashed boxes are <b>facts still needed from the client</b>. Nothing has been invented.
</div>

<script>
(function(){{
  var routes = {json.dumps(list(pages.keys()))};
  var LANDING = {json.dumps(default_route)};
  var HOME = routes.indexOf(LANDING) >= 0 ? LANDING : routes[0];
  function norm(h){{
    h = (h || '').replace(/^#/, '');
    if (!h) return HOME;
    if (!h.startsWith('/')) h = '/' + h;
    if (h !== '/' && !h.endsWith('/')) h += '/';
    return routes.indexOf(h) >= 0 ? h : HOME;
  }}
  function show(r){{
    document.querySelectorAll('.rt').forEach(function(el){{
      el.hidden = el.getAttribute('data-route') !== r;
    }});
    document.querySelectorAll('.rt:not([hidden]) [aria-current]').forEach(function(el){{
      el.removeAttribute('aria-current');
    }});
    window.scrollTo(0,0);
    reveal();
  }}
  function reveal(){{
    document.querySelectorAll('.rt:not([hidden]) .reveal').forEach(function(el){{
      el.classList.add('is-in');
    }});
  }}
  // Rewrite internal links to hash routes
  document.addEventListener('click', function(e){{
    var a = e.target.closest && e.target.closest('a');
    if (!a) return;
    var href = a.getAttribute('href') || '';
    if (/^(https?:|tel:|mailto:|#)/.test(href)) return;
    e.preventDefault();
    location.hash = href;
  }});
  window.addEventListener('hashchange', function(){{ show(norm(location.hash)); }});
  // Mobile nav
  document.addEventListener('click', function(e){{
    var t = e.target.closest && e.target.closest('.nav-toggle');
    if (!t) return;
    var nav = t.parentElement.querySelector('.nav');
    if (nav) nav.classList.toggle('is-open');
  }});
  var hdr = document.querySelectorAll('.header');
  window.addEventListener('scroll', function(){{
    hdr.forEach(function(h){{ h.classList.toggle('is-scrolled', window.scrollY > 8); }});
  }}, {{passive:true}});
  show(norm(location.hash));
}})();
</script>
"""
dest = pathlib.Path(argv[0]) if argv else pathlib.Path(f"/tmp/{pathlib.Path.cwd().name}-preview.html")
dest.write_text(out)
print("written to", dest)
print(f"{len(pages)} routes bundled -> {len(out)//1024} KB")
print("routes:", ", ".join(list(pages.keys())[:8]), "...")

File: clients/loxley-solar/bundle-preview.mjs
// Bundle dist/ into one self-contained preview HTML for the temp share link.
// No iframes: the artifact host sizes pages by content height, so the site
// pages are swapped directly into the document flow and their inline scripts
// re-executed on each switch. Fonts/images are deduped data-URI tokens.
import { readFileSync, writeFileSync, existsSync } from 'node:fs';
import { resolve } from 'node:path';

const DIST = resolve('dist');
const OUT = process.argv[2] || '/tmp/preview.html';

const SITE = {
  title: 'Loxley Solar — temp preview',
  brand: 'Loxley Solar',
  note: 'Demo data marked TEMP swaps before go-live. Estimator maths uncalibrated; form runs in email mode.',
  bar: { ground: '#0f3b2e', surface: '#14462f', edge: '#275940', ink: '#f2f7f1', soft: '#b9ccbe', accent: '#f5b301', onAccent: '#1a2420' },
};
const ROUTES = [
  { key: 'home', label: 'Home', path: '/', file: 'index.html' },
  { key: 'services', label: 'Services', path: '/services/', file: 'services/index.html' },
  { key: 'about', label: 'About', path: '/about/', file: 'about/index.html' },
  { key: 'blog', label: 'Blog', path: '/blog/', file: 'blog/index.html' },
  { key: 'post', label: 'Solar costs', path: '/blog/solar-panel-cost-sheffield/', file: 'blog/solar-panel-cost-sheffield/index.html' },
  { key: 'notfound', label: '404', path: '/404.html', file: '404.html' },
];

const MIME = { woff2: 'font/woff2', webp: 'image/webp', svg: 'image/svg+xml', png: 'image/png' };
const assets = {};
let tokenN = 0;
const tokenFor = {};
function assetToken(path) {
  if (tokenFor[path]) return tokenFor[path];
  const f = resolve(DIST, '.' + path);
  if (!existsSync(f)) throw new Error('asset missing: ' + path);
  const ext = path.split('.').pop();
  const tok = `@@A${tokenN++}@@`;
  assets[tok] = `data:${MIME[ext]};base64,${readFileSync(f).toString('base64')}`;
  tokenFor[path] = tok;
  return tok;
}

const pages = {};
for (const r of ROUTES) {
  let html = readFileSync(resolve(DIST, r.file), 'utf8');
  html = html.replace(/<link rel="preload"[^>]*\/?>/g, '');
  html = html.replace(/srcset="([^"]+)"\s*sizes="[^"]*"/g, (m, set) => {
    const last = set.split(',').pop().trim().split(/\s+/)[0];
    return `data-collapsed-src="${last}"`;
  });
  html = html.replace(/src="(\/[^"]+\.(?:webp|svg|png))"/g, (m, p) => `src="${assetToken(p)}"`);
  html = html.replace(/data-collapsed-src="(\/[^"]+)"/g, (m, p) => { assetToken(p); return `src="${tokenFor[p]}"`; });
  html = html.replace(/url\('?(\/fonts\/[^')]+)'?\)/g, (m, p) => `url('${assetToken(p)}')`);
  // head styles carry the page CSS (inlineStylesheets: always)
  const head = html.slice(0, html.indexOf('</head>'));
  const css = [...head.matchAll(/<style>([\s\S]*?)<\/style>/g)].map(m => m[1]).join('\n');
  const body = html.slice(html.indexOf('<body>') + 6, html.lastIndexOf('</body>'));
  pages[r.path] = { css, body };
}

const shell = `<title>${SITE.title}</title>
<style>
  body { margin: 0; padding: 0; }
  .pv-bar {
    display: flex; align-items: center; gap: 14px; flex-wrap: wrap;
    padding: 10px 16px;
    background: ${SITE.bar.ground}; color: ${SITE.bar.ink};
    border-bottom: 1px solid ${SITE.bar.edge};
    font: 500 14px/1.4 system-ui, sans-serif;
  }
  .pv-bar .brand { font-weight: 700; letter-spacing: -.01em; }
  .pv-bar .brand small { color: ${SITE.bar.soft}; font-weight: 500; margin-left: 8px; }
  .pv-bar nav { display: flex; gap: 4px; flex-wrap: wrap; margin-left: auto; }
  .pv-bar button {
    font: 600 13px system-ui, sans-serif; color: ${SITE.bar.soft};
    background: transparent; border: 1px solid transparent; border-radius: 999px;
    padding: 7px 14px; cursor: pointer;
  }
  .pv-bar button:hover { color: ${SITE.bar.ink}; border-color: ${SITE.bar.edge}; }
  .pv-bar button[aria-current="true"] { color: ${SITE.bar.onAccent}; background: ${SITE.bar.accent}; }
  .pv-bar button:focus-visible { outline: 2px solid ${SITE.bar.accent}; outline-offset: 2px; }
  .pv-bar .note { color: ${SITE.bar.soft}; font-size: 12px; width: 100%; }
  @media (min-width: 900px) { .pv-bar .note { width: auto; } }
</style>
<style id="pv-page-css"></style>
<div class="pv-bar">
  <span class="brand">${SITE.brand}<small>temp preview · Stacked Out</small></span>
  <nav id="pv-tabs" aria-label="Preview pages"></nav>
  <span class="note">${SITE.note}</span>
</div>
<div id="pv-stage"></div>
<script>
  var ASSETS = ${JSON.stringify(assets).replace(/</g, '\\u003c')};
  var PAGES = ${JSON.stringify(pages).replace(/</g, '\\u003c')};
  var ROUTES = ${JSON.stringify(ROUTES.map(({ key, label, path }) => ({ key, label, path })))};
  var stage = document.getElementById('pv-stage');
  var pageCss = document.getElementById('pv-page-css');
  var tabs = document.getElementById('pv-tabs');

  function hydrate(s) { return s.replace(/@@A\\d+@@/g, function (t) { return ASSETS[t]; }); }

  function show(path) {
    var anchor = '';
    if (path.indexOf('#') !== -1) {
      anchor = path.slice(path.indexOf('#') + 1);
      path = path.slice(0, path.indexOf('#')) || '/';
    }
    var route = ROUTES.find(function (r) { return r.path === path; }) || ROUTES[0];
    var page = PAGES[route.path];
    pageCss.textContent = hydrate(page.css);
    stage.innerHTML = hydrate(page.body);
    // innerHTML never runs scripts: recreate them so the page behaves.
    stage.querySelectorAll('script').forEach(function (old) {
      var s = document.createElement('script');
      s.textContent = old.textContent;
      old.replaceWith(s);
    });
    tabs.querySelectorAll('button').forEach(function (b) {
      b.setAttribute('aria-current', String(b.dataset.path === route.path));
    });
    if (anchor) {
      var t = document.getElementById(anchor);
      if (t) t.scrollIntoView();
    } else {
      window.scrollTo(0, 0);
    }
  }

  ROUTES.forEach(function (r) {
    var b = document.createElement('button');
    b.textContent = r.label; b.dataset.path = r.path;
    b.addEventListener('click', function () { show(r.path); });
    tabs.appendChild(b);
  });

  // Internal site links switch pages; anchors, tel: and mailto: stay native.
  stage.addEventListener('click', function (e) {
    var a = e.target.closest ? e.target.closest('a') : null;
    if (!a) return;
    var href = a.getAttribute('href') || '';
    if (href.charAt(0) === '/') { e.preventDefault(); show(href); }
  }, true);

  show('/');
</script>`;

writeFileSync(OUT, shell);
console.log('wrote', OUT, Math.round(shell.length / 1024) + 'KB,', Object.keys(assets).length, 'assets');
